docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,220 @@
+In practice, this assumption is not restrictive when doing optimization as such problem require a real-valued objective (as there is no natural ordering of the complex numbers).
+
+Under this assumption, we can show that :math:`W = CW^*` (we use :math:`*` to denote complex conjugation here) and so only one of the two values actually need to be "backwarded through the graph" as the other one can easily be recovered.
+To simplify internal computations, PyTorch uses :math:`2 * CW` as the value it backwards and returns when the user asks for gradients.
+Similarly to the real case, when the output is actually in :math:`\mathcal{R}^M`, backward mode AD does not compute :math:`2 * CW` but only :math:`v^T (2 * CW)` for a given vector :math:`v \in \mathcal{R}^M`.
+
+For forward mode AD, we use a similar logic, in this case, assuming that the function is part of a larger function whose input is in :math:`\mathcal{R}`. Under this assumption, we can make similar claim that every intermediary result corresponds to a function whose input is in :math:`\mathcal{R}` and in this case, we can show that :math:`W = CW` for the intermediary functions.
+To make sure the forward and backward mode compute the same quantities, the forward mode also computes :math:`2 * CW`.
+Similarly to the real case, when the input is actually in :math:`\mathcal{R}^N`, forward mode AD does not compute :math:`2 * CW` but only :math:`(2 * CW) u` for a given vector :math:`u \in \mathcal{R}^N`.
+
+
+Current backward mode gradcheck behavior
+----------------------------------------
+
+Functions in :math:`\mathcal{R}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We want to test a function :math:`f: \mathcal{R}^N \to \mathcal{R}^M, x \to y`.
+To do so, we reconstruct the full Jacobian matrix :math:`J_f` of size :math:`M \times N` in two ways.
+The analytical version uses our backward mode AD while the numerical version uses finite difference.
+We then compare each element of the two reconstructed Jacobian matrices to ensure they do match.
+
+Slow real input numerical evaluation
+""""""""""""""""""""""""""""""""""""
+
+If we consider the most basic case of a one-dimensional function (:math:`N = M = 1`), then we can use the basic finite difference formula from `the wikipedia article <https://en.wikipedia.org/wiki/Finite_difference>`_ (we use the central difference for better numerical properties):
+
+.. math::
+    \frac{\partial y}{\partial x} \approx \frac{f(x + eps) - f(x - eps)}{2 * eps}
+
+This formula easily generalize for multiple outputs (:math:`M \gt 1`) by having :math:`\frac{\partial y}{\partial x}` be a column vector of size :math:`M \times 1` like :math:`f(x + eps)`. In that case, the above formula can be re-used as-is and approximate the full Jacobian matrix with only two evaluations of the user function.
```

Comment:
```suggestion
This formula can be easily generalized for multiple outputs (:math:`M \gt 1`) by having :math:`\frac{\partial y}{\partial x}` to be a column vector of size :math:`M \times 1` like :math:`f(x + eps)`. In that case, the above formula can be re-used as-is and the full Jacobian matrix can be approximated with only two evaluations of the user function.
```

Reply:
Added a comment for the function.
I did not change the tenses though to remain consistent with the rest of the note that does not (should not?) be using passive tenses.